The repeater!

jneen opened the pull request:
This pull request includes the repeater, as well as the refactoring necessary to make it work.

Why do we need to refactor like this? The TLDR is: we need to be able to access the data from helpers, which means it needs to be registered on `res.locals` instead of passed into `res.render(...)`.

jneen commented:
Breaking changes include:

* rendering is no longer done with `res.render(route.name, getViewData(...))`. Instead, we use the `route.loadKeys(...)` or `route.loadSchema(Schema)` to load data from the session, and the middleware `route.render()` to render the template. This ensures that all data is available on `res.locals` and can be seen by helpers. `getViewData` et al have been removed.

* the argument order of `radioButtons` and `checkBoxes` have changed, and errors are no longer manually passed in. This makes them consistent with the other controls, and usable inside repeaters. The new order is:

``` jinja
{{ checkBoxes('my_field', 'question.label', { 'val1': 'label1', 'val2': 'label2 }) }}
{{ radioButtons('my_field', 'question.label', { 'val1': 'label1', 'val2': 'label2' }) }}
```

The `checkBoxes` control will result in a value that is an object with the selected keys set to the string "on". For example, if I check the box next to 'label1', the backend will receive `my_field: { val1: 'on' }`. The `radioButtons` control will result in a string indicating which value was selected - if I select 'label1', the backend will receive `my_field: 'val1'`.

jneen commented:
Because it was easier to include than not, I've also included our `radioBlock` macro, which I'll add an example for. This is intended for radio buttons that select segments of a form, for example entering height in imperial or metric units:

``` jinja
        {% call radioBlock('heightSystem', null,
                           { 'metric': 'height.enter-centimetres' }) %}
            <div id="metric" style="display: {{displayMetric}};">
                <div class="w-3-4">
                    {{ textInput('cm', 'height.centimetres', { class: 'w-1-8' }) }}
                </div>
            </div>
        {% endcall %}

        {% call radioBlock('heightSystem', null,
                           { 'imperial': 'height.enter-feet-inches' }) %}
            <div id="imperial" style="display: {{displayImperial}};">
                <div class="w-1-8">
                    {{ textInput('feet', 'height.feet', { class: 'w-3-4' }) }}
                    {{ textInput('inches', 'height.inches', { class: 'w-3-4' }) }}
                </div>
            </div>
        {% endcall %}
```

timarney commented on public/js/repeater.js:
```diff
@@ -0,0 +1,185 @@
+    return out
+  }
+
+  function Block() { this.init.apply(this, arguments) }
```
Let's add comments here - to say what a block is.

timarney commented on public/js/repeater.js:
```diff
@@ -0,0 +1,185 @@
+      // we use one global listener for removal, because the repeat links may
+      // get added and removed from the DOM, and this means we don't have to
+      // re-register the event listeners when we clone the nodes.
+      this.container.addEventListener('click', function(evt) {
```
💯

timarney commented on public/js/repeater.js:
```diff
@@ -0,0 +1,185 @@
+    }
+
+    _.repeat = function() {
+      if (!this.instances.length) throw new Error('empty instances, can\'t repeat!')
```
Would prefer to use template literals vs escaping stings `empty instances, can't repeat!`

timarney commented on public/js/repeater.js:
```diff
@@ -0,0 +1,185 @@
+    // private functions
+    function reindex(str, index) {
+      // it's always going to be the first [0] or [1] or etc.
+      return str.replace(/\[\d+\]/, '['+index+']')
```
Again here `[${index}])` would prefer template literals

jneen replied:
oh! I should say, this is front-end javascript, run on the browser, so I'm not sure what the compatibility story is for these kinds of features. Can we rely on template literals in our users' browsers?

timarney replied:
No worries Webpack will take care of this to make it backwards compat to ES5  or whatever we want to target.

timarney commented on utils/context.helpers.js:
```diff
@@ -0,0 +1,59 @@
+  // we would ideally use a caller() macro for this, but those are...
+  // unreliable at best.
+  const keyPath = []
+  res.locals.enterContext = (key) => { keyPath.push(key) }
```
I won't have a chance to dive into this until later in the week - what's going on with enter and exit content here?

jneen replied:
I suggest you have a look at the tests, I tried to make it clear there :]. Basically, they maintain this key path, which affects the "context" for `getData`, `getError`, and `getName`. This way, we can push and pop to enter and exit certain branches of the data tree. This is important because now we can have nested data in the session / body, and we need a generic way to traverse it that integrates with our form controls.

timarney replied:
Okay will take a look when I can

timarney commented on utils/context.helpers.js:
```diff
@@ -0,0 +1,59 @@
+  res.locals.getData = (...keys) => lookupKeypath(res.locals, keyPath.concat(keys))
+  res.locals.getError = (...keys) => (res.locals.errors || {})[errorPath(keyPath.concat(keys))]
+
+  res.locals.pad = (arr, len) => {
```
Think we should be able to add default args here (could be wrong as I haven't dived into the code)

res.locals.pad = (arr=[], len=1)

jneen replied:
Default args are great for when the caller doesn't pass arguments in - but in nunjucks it's very common to explicitly pass `undefined` around.

jneen commented on public/js/repeater.js:
```diff
@@ -0,0 +1,185 @@
+  var repeatedSets
+
+  // Since we can't depend on Array.prototype.map
+  function map(arr, fn) {
```
Oh neat! Does that get us ES.next features as well?

timarney replied:
Yes.

You can use whatever you need and we'll gets things setup to ensure it compiles back.

jneen replied:
Okay! How about we get that set up first, so I can test it?

timarney replied:
If you look at the fileInput in the utils ... the fileInput gets exported

```javascript
export const fileInput = () => {
 ....
```

Than in the personal route in the js dir it's imported

```javascript
import { fileInput } from '../../../utils/fileInput'
import { registerToggle } from '../../../utils/toggleClass'
;(function(document, window, index) {
  fileInput()
  registerToggle('.multiple-choice__item input', '.notify-type-toggle-container')
})(document, window, 0)

```

In the Wepack config you need to register the file under the entry "key"

```
const config = getConfig({
    mode: argv.mode,
    entry: {
      styles: './assets/scss/app.scss',
      personal: './routes/personal/js/personal.js',
      addresses: 'your-path-here'
    },
```

In the controller if you use 
```
const js = getClientJs(req, route.name)
```

That code will take care of the path 

And you pass it to the view

```
routeUtils.getViewData(req, { jsFiles: js ? [js] : false }),
```

timarney replied:
If we need to target or add things to Webpack that can be done here https://github.com/cds-snc/webpack-starter

Or the config can be modified locally using Webpack merge.

jneen replied:
I'm *very* confused about this bit - why is the webpack stuff a separate repository? How is it configured?

jneen replied:
(also seeing as we're deprecating `getViewData` we'll need another way of passing that)

timarney replied:
> I'm _very_ confused about this bit - why is the webpack stuff a separate repository? How is it configured?

In general devs don't "enjoy" or know how to configure Webpack so we're doing it for them and allowing them them to override.  Most projects will need roughly the same config so much like the starter nice to give them a base setup.  Projects such as Vue CLI, Next JS, Razzle and a bunch of others follow this pattern.  

On top of this again like this repo we can manage the dependancies on layer up - so bump once on X packages upstream and the end user just needs to bump one package.  IMO better to do that work up a level if we can.

As far as config:  

If you look at the Webpack file in this repo

https://github.com/cds-snc/node-starter-app/blob/master/webpack.config.js

The other repo gives you some config defaults 

```
const path = require('path')

module.exports = (env, argv) => {
  const { getConfig } = require('@cdssnc/webpack-starter')
  const config = getConfig({
    ...
  return config
}
```

And webpack merge can be used to extend that config or the user can opt out and just use there own config.

timarney replied:
> (also seeing as we're deprecating `getViewData` we'll need another way of passing that)

Understand about deprecating `getViewData`  (or is it being removed) but not sure what the question is here.

I can lend a hand getting the Webpack stuff going if you like.

jneen replied:
I see! So if I understand right, the repo there is just a dummy package used to pull in dependencies.

Not a question, just adding a TODO for myself, because the way we handle `jsFiles` needs to be thought out a little more with the new way of rendering. Perhaps it can be rolled into `route.render()`.

timarney replied:
Right it gives a based setup and essentially stops the duplication of have all those deps in all the repos.

Also note:  In that repo I wrote a custom Webpack plugin that writes a "_filelist.json" file

Which is just a lookup file.


```_filelist.json
{
	"book":"book.930291000905c8c00779.js"
}
```

To avoid cache issues we have Wepack setup to add a hash to the end of the filename
```
 output: {
      filename: '[name].[chunkhash].js',
```

So technically for each route we don't really know what the file name will be.  We look it up using the _filelist.json.  That part all works fine just wanted to mention the reason it's there.

At some point I want to get into hot reloading js file (replacing them in page vs nodemon full page reload) but that's down the road.

jneen commented on public/js/repeater.js:
```diff
@@ -0,0 +1,185 @@
+window.Repeater = (function() {
```
Also cool - since we didn't have webpack in the passport app, I was using old-style modularization, where we expose a single global variable per module. But you're right, now that we have webpack, that's the right way to do it :]

jneen commented:
@timarney Those are all very valid concerns - I was also a bit worried when I was implementing these. What I eventually came to, though, was that the current API was inconsistent and brittle enough that it was probably best to fix them while we're still in early adoption. After this, we should be able to freely change how things work without breaking user code.

I agree that `enterContext` and `exitContext` can be a bit tough to read for those who aren't used to working with dynamic contexts - but the alternative is to place the burden of tracking names, errors, and values on the form designer, which I think is a non-starter, especially since they all have slightly different APIs. I should note that those two functions should *only* be used by those implementing form controls that use nested content (like the repeater), and never by a form designer, or those implementing simple controls. For those implementing simple controls, `getData`, `getName`, and `getError` are available.

timarney commented:
@jneen can you ping me when it's a good time to pull down and review?

I currently get an error which assuming is given this is a work in progress.
```
express-validator: a validator/sanitizer with name default does not exist
Listening on port 3000 production
contextMiddleware
contextMiddleware
☠️ Error => Cannot read property 'msg' of undefined
```

timarney commented:
> * the argument order of `radioButtons` and `checkBoxes` have changed, and errors are no longer manually passed in. This makes them consistent with the other controls, and usable inside repeaters. The new order is:

Given this changes the argument order I would like to take this opportunity to to be able to pass specific attributes to to each radio / checkbox.

I started down that road in a branch called toggle but using the current field order but opting to hold on that work until this is in.

**Passing attrs to radios**

```nunjucks
// note the <input {{inputAttr}} id="{{ key }} ....

<div class="multiple-choice multiple-choice--radios" id="{{ key }}">
                {% if errors and errors[key] %}
                    {{ validationMessage(errors[key].msg, key) }}
                {% endif %}
                {% for index, val in values %}
                    {% set inputAttr = attributes[index] %}
                    <div class="multiple-choice__item">
                        <input {{inputAttr}} id="{{ key }}{{ val }}" name="{{ key }}" type="radio" value="{{ val }}" {% if value == val %} checked="checked" {% endif %} {% if errors and errors[key] %} aria-describedby="{{ key + '-error' }}" aria-invalid="true" {% endif %}>
                        <label for="{{ key }}{{ val }}">{{ __(val) }}</label>
                    </div>
                {% endfor %}
            </div>
```

<hr>

**View**
```nunjucks
// note the {% set attrs = 
<div class="notify-type-toggle-container hide">
            {% set attrs = {hint: 'form.send_notifications.desc', yes: "data-toggle=on", no: "data-toggle=off"} %}
            {% set options =  { 'yes':'Yes','no':'No'} %}
            {{ radioButtons('send_notifications',options, data.send_notifications, 'form.send_notifications', errors, attrs, { required: true }) }}
            <div class="toggle">
                {{ checkBoxes('notify_type', {'sms':'SMS','email':'Email'}, data.notify_type, 'form.notify_by', errors, {hint: 'form.notify_by.desc'} ) }}
            </div>
        </div>
```

With the result being something like this.
<img width="851" alt="Screen Shot 2019-11-26 at 7 56 51 AM" src="https://user-images.githubusercontent.com/62242/69635434-5ccb9600-1022-11ea-8e4c-b27546b4bde4.png">

This idea being if I need to add "random" attrs to a radio we can do that.

That goes back to this issue when I was thinking of this type of format

**Passing attributes to single radios / checkboxes**

https://github.com/cds-snc/node-starter-app/pull/42#issuecomment-534096584

```
{{ radioButtons('group_name_for_input', { 'input_1': {classes: "yyy" , id: '"yyy", data-attr: "yyy"},'input_2':{classes: "zzz" , id: '"zzz", data-attr: "zzz"}}) }}
```
*Note:*Given what we know now I'm sure we can come up with a good format




**Nunjucks spread params filter**

With that said:

My goal was to end up using a spread params filter which would put the end user in control of all the fields without us needing to define.  Perhaps we can utilize this at least in part.

https://github.com/cds-snc/node-starter-app/pull/69#issue-321242317

Happy to discuss some options / ideas ... overall looking for ways to not need to modify the macros when an extra field is needed.

jneen commented:
Sounds good, i have a few ideas.

timarney commented:
Also FYI - 

Let's plan for a major 7.0 release.  We can setup a branch.  

We would roll out it out sometime after the meeting that Ross scheduled (pending outcomes). Map out a roadmap + features for a major breaking release. 

Moving forward -> setup a request for commit proposal system.

cc: @dsamojlenko

jneen commented:
@timarney would it be alright to do the radio button attributes bit in a separate PR? this one is getting pretty big as it is.

timarney commented:
> alright to do the radio button attributes bit in a separate PR

For sure was just flagging that use case

jneen commented:
@timarney should be green, ready for review. if we want to merge this to a special 7.0 branch that'd be fine too :]

timarney commented:
@JuliannaR can you take a look https://cds-node-starter-pr-132.herokuapp.com/en/addresses

dsamojlenko commented:
Late to the party here... but wondering if we can implement this in a way where the Nunjucks templates are decoupled from the framework. ie, all necessary attributes are passed into the macro, and don’t require access to functions injected by middleware, or knowledge of where data is coming from (ie, the current iteration has references out to `data.[fieldname]`, or how the macros all know about the `errors` object)

Perhaps there could be a set of dumb core macros for the individual field types that are wrapped/extended by macros that enable this advanced integration with the framework.

This way the underlying macros would be more portable/reusable in other contexts.

Don't mean to throw a monkey wrench in here... just think this would help with the progression stuff we were talking about earlier. Might also help limiting the breakingness of this major breaking change.

timarney commented:
@jneen

Noting some things now as it's the first time we've had access to a preview link (as of yesterday).

## Accessibility 
@JuliannaR 

First thing I found on the a11y front.  If you tab to add new address and then tab again to reach the first field in the newly added "sub-form" it hides (the newly created fields) and jumps to the continue button. 
 The fields are added but I could access them and ended up with a bunch of address fields when I reverted to my trackpad + clicking.

The addresses remove fields links don't have unique identifiers .... and I think we'll uncover a bunch more as we go.

## Usability Concerns

The above (which I know we haven't spent time on) has me questioning the usability in general of the repeater pattern (not the code implementation).  Understanding the need we see based on the 2026 app and I'm sure most of the Gov pdf forms but does this translate well for accessible forms?  Have we looked at others using this pattern?

## Research & Design

Was user research done on the 2026 app that would indicate this is a good pattern to move forward with?

#Also flagging this link for @MithulaNaik https://cds-node-starter-pr-132.herokuapp.com/en/addresses

<hr> 

Appreciate all the work and effort to date, just wanting to make sure we cover all our bases that we want to move forward with this 100%. 

cc: @rossferg @cgovias 

Also flagging @dsamojlenko comments above so they don't get buried.

> macros decoupled from the framework 

Would love to see if there's any way to pull this off it would be really nice to be able to use the macros elsewhere if possible i.e. https://www.11ty.io supports Nunjucks and might be a good fit for us to use for static sites (use our macros and styles).

jneen commented:
@dsamojlenko Hi! Thanks for taking the time to look at this. My first attempt at building something like this used only nunjucks macros. Unfortunately, nunjucks is *extremely* restrictive in its scoping rules, to the point that `call` macros do not even retain access to their containing scope. There is a chance that in the future they will allow dynamic or global variables, and we will be able to refactor this into pure nunjucks. However, for the time being, the only way to consistently change context is to use helper functions injected in a middleware.

jneen commented:
If we want to do research on this sort of thing, I would suggest we put some design resources into it! As far as it goes now we are only generating `fieldset` tags with no border, and everything seems to run together. I will take a look at the tabbing issue, hopefully get that fixed, and I should be able to add a unique id to the remove links.

The use case here, to be clear, is when the number of user inputs is potentially unbounded - for example, "all addresses in the last two years", where someone could have had a very large amount of addresses. There may be other ways to solve this problem, but the general "add another" functionality is definitely needed (compare to existing forms, where applicants are asked to design their own attachment format "if there isn't room").

jneen commented:
cc @cgovias re: the user research we did in 2620 - I was not able to come as a distributed employee :]

jneen commented:
Also @dsamojlenko requiring the form designer to pass in all necessary attributes was something we considered, but I ultimately decided that it unnecessarily increased the complexity burden in something that was going to be repeated many times, and introduced a large possibility of very subtle coding errors. Consider that every field would need to receive not only the current value, but also the error (if present), and a scoped name - and all three of these have subtly different access patterns! For the scoped name, we need `addresses[0][attr]`, for the error `addresses[0].attr`, and for the value we have to iterate into an actual data structure. This is *much* easier done behind the scenes.

jneen commented:
Nevermind, I found the research document @timarney @JuliannaR https://docs.google.com/document/d/11ZsxDD2twtmwvrIycEJtd2XSF6a0eGnxAwA6HO_XyzM/edit#heading=h.bsucfu4hx5bt

It touches on a few bugs and... the fact that there's no border between fieldsets.

JuliannaR commented:
@timarney yes, the remove and add don't unique identifiers so hard to tell which you are removing or adding. 

If you hit enter after marking any checkbox it advances you to the next page for with the confirmation of addresses.
Spacing is difficult to read/understand that you are interacting with different segments to enter content.

<img width="1156" alt="Screen Shot 2019-12-04 at 10 45 52 AM" src="https://user-images.githubusercontent.com/35537308/70157732-17434480-1684-11ea-9f9a-3d68ac82b490.png">

Also, unlikely to translate into accessible forms. Some testing will have to be looked at when it gets there.

JuliannaR commented:
@timarney yes, the remove and add don't unique identifiers so hard to tell which you are removing or adding. 

If you hit enter after marking any checkbox it advances you to the next page for with the confirmation of addresses.
Spacing is difficult to read/understand that you are interacting with different segments to enter content.

<img width="1156" alt="Screen Shot 2019-12-04 at 10 45 52 AM" src="https://user-images.githubusercontent.com/35537308/70157732-17434480-1684-11ea-9f9a-3d68ac82b490.png">

Also, unlikely to translate into accessible forms. Some testing will have to be looked at when it gets there.

jneen commented:
@JuliannaR I can add an `id` attribute to all of those - are there other attributes I should add to things to link them?

jneen commented:
Also, submitting the form on `<enter>` seems to be the default behaviour, and is the same on any textbox. Should we override this with javascript? And if so, what behaviour would be expected?

jneen commented:
I have added unique identifiers to the repeat and remove-repeat links - the limitation now is there can't be more than one of them (but i don't think that's actually a use-case).

@JuliannaR I did some navigating around with the keyboard and you're absolutely right, the links are pretty hard to use. I've added some javascript code that allows pressing the spacebar to activate them, but I'm not sure this is the correct approach. Would a `<button>` be more appropriate here?

jneen commented:
I have significantly improved keyboard navigation and redone the js includes to work with the new system. Any `client.js` file in a route directory will be automatically compiled by webpack and included in the page via base.njk, and `global/client.js` will be included in every page. Shared code is kept in `assets/js`.

jneen commented:
rebased against master.

jneen commented:
The LGTM alert is for the `document.write` in the polyfill script, which was already present, I just moved it :]

jneen commented:
This will 100% need design work, but I don't know how to allocate that. @rossferg, thoughts?

rossferg commented:
> This will 100% need design work, but I don't know how to allocate that. @rossferg, thoughts?

Noted. Will find design support.

jneen commented:
Just so everyone knows I'll be out on family-related leave for a good while starting tomorrow, so today is the last day I can shepherd this.

jneen commented:
cc @timarney just bumping this because it's the last day I can do any fixes!

jneen commented:
I should say, another benefit of using `getData`, `getError`, and `getName` is that these are all fully abstract, so they should actually be easier to use in different contexts. If we are using this in a static page, for example, there is no session integration, so we can just return `null` from `getError` and `getData`.

jneen commented:
The LGTM bot warning is buggy - if you click through you will see there are 0 alerts.

jneen commented:
Thank you! I agree, the RFC process would have helped this significantly.

As for the "coupling" of the macros, do you mean coupling them to the specific implementation of the starter-repo? I would actually argue that this approach is more decoupled, as any system can define `getName`, `getError`, and `getData` and use these immediately (static site generators, for example), rather than having all that logic within handlebars code.

jneen commented:
One last question @timarney: should this be merged to master given that it is a major breaking change?

jneen commented:
I also agree that this PR grew a bit too large. I likely would have added the repeater functionality in a separate branch, but I had been under the impression that it was important to get this through before my family leave based on discussions with @rossferg. We had also already finished the coding in the 2620 branch, and this was a simple port.

Many of the changes here were me cleaning up as I went, attempting to make the codebase nicer than how I found it - especially as some of those design decisions that had been made in haste blocked the way forward.

jneen commented:
Okay! Without further input, I think it's best to merge to master so we can move forward from here. Thanks!
